File: lib/signup.js
'use strict'

import delay from 'delay'
import apiClient from '../util/api-client.js'

let message = ''
let timeLeft = ''
let status = 1

export const signup = async (browser, user, profileId) => {
  const page = await browser.newPage()
  await page.goto('https://instagram.com/accounts/emailsignup/')

  await page.waitForSelector('input[name=emailOrPhone]', { visible: true })
  await delay(300)
  await page.type('input[name=emailOrPhone]', user.number, { delay: 27 })

  await delay(500)
  await page.type('input[name=fullName]', `${user.firstName} ${user.lastName}`, { delay: 82 })

  await delay(700)
  const username = await page.$eval('input[name=username]', (el) => el.value)
  if (username) {
    user.username = username
  } else {
    await page.type('input[name=username]', user.username, { delay: 67 })
  }
  await page.keyboard.press('Tab')
  await delay(1000)
  await page.keyboard.press('Space')
  await delay(300)

  await delay(300)
  await page.type('input[name=password]', user.password, { delay: 42 })

  await delay(700)
  const [signupButton] = await page.$x('//button[@type="submit"]')
  user.username = await page.$eval('input[name=username]', (el) => el.value)

  await delay(4000)

  const addAccountDataToProfile = async (account) => {
    await delay(4000)
    const url = `http://localhost:3001/profile/${profileId}/addAccount`
    /*    const headers = {
          Accept: '*!/!*',
          'Content-Type': 'application/json'
        } */
    const accountDetails = {
      account: {
        username: user.username,
        password: user.password,
        phoneNumber: user.number
      }
    }
    try {
      const response = await apiClient.post(url, accountDetails)

      // Process the response data
      console.log('Account added to profile', response)
    } catch (error) {
      console.error('Failed to add account to profile:', error)
    }
  }

  // OTP message Method
  const fetchOTPDetails = async () => {
    while (![0, 2, 3, 4, 5, 6].includes(status)) {
      try {
        delay(3000)
        const SmSPoolFetchAPI = 'https://api.smspool.net/sms/check'
        const response = await fetch(`${SmSPoolFetchAPI}?orderid=${user.orderId}&key=${user.key}`)
        const jsonData = await response.json()
        message = String(jsonData.sms)
        status = jsonData.status
        timeLeft = jsonData.time_left
        console.log('[OTP Response]', { jsonData, message, timeLeft })
        await new Promise(resolve => setTimeout(resolve, 4000))
      } catch (error) {
        console.error('An error occurred while checking API status:', error)
        break
      }
    }
    console.log('[Status of the API is]', status)
  }

  await Promise.all([signupButton.click({ delay: 70 })])

  await delay(5000)

  function getRandomInteger (min, max) { // min and max are exclusive
    return String(Math.floor(Math.random() * (max - min)) + min)
  }

  await page.select('select[title="Month:"]', getRandomInteger(0, 12))
  await page.select('select[title="Day:"]', getRandomInteger(0, 29))
  await page.select('select[title="Year:"]', getRandomInteger(1965, 2000))
  await delay(5000)
  const nextButton = await page.$x('//button[@type="button"]')
  if (nextButton.length > 0) {
    nextButton[1].click()
  }
  await fetchOTPDetails()
  await page.waitForSelector('input[name=confirmationCode]', { visible: true })
  await delay(300)
  if (status !== 1) {
    await page.type('input[name=confirmationCode]', message, { delay: 27 })
    await delay(5000)
    const confirmationSignupButton = await page.$x('//button[@type="button"]')
    if (confirmationSignupButton.length > 0) {
      await confirmationSignupButton[0].click()
      await addAccountDataToProfile()
    }
  }
  await page.close()
}

File: lib/signup.py
import time
import random
import requests
from selenium import webdriver
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.support.ui import Select

def signup(browser, user):
    browser.get('https://instagram.com/accounts/emailsignup/')
    time.sleep(0.3)

    email_input = browser.find_element_by_css_selector('input[name=emailOrPhone]')
    email_input.send_keys(user['number'])
    time.sleep(0.5)

    full_name_input = browser.find_element_by_css_selector('input[name=fullName]')
    full_name_input.send_keys(user['firstName'] + ' ' + user['lastName'])
    print("Hello")
    time.sleep(0.7)

    username_input = browser.find_element_by_css_selector('input[name=username]')
    username = username_input.get_attribute('value')
    if username:
        user['username'] = username
    else:
        username_input.send_keys(user['username'])

    username_input.send_keys(Keys.TAB)
    time.sleep(1)
    browser.find_element_by_tag_name('body').send_keys(Keys.SPACE)
    time.sleep(0.3)

    password_input = browser.find_element_by_css_selector('input[name=password]')
    password_input.send_keys(user['password'])
    time.sleep(0.7)

    signup_button = browser.find_element_by_xpath('//button[@type="submit"]')
    user['username'] = browser.find_element_by_css_selector('input[name=username]').get_attribute('value')
    time.sleep(4)

    def fetchOTPDetails():
        status = 1
        while status not in [0, 2, 3, 4, 5, 6]:
            try:
                time.sleep(3)
                SmSPoolFetchAPI = 'https://api.smspool.net/sms/check'
                response = requests.get(f'{SmSPoolFetchAPI}?orderid={user["orderId"]}&key={user["key"]}')
                jsonData = response.json()
                message = str(jsonData['sms'])
                status = jsonData['status']
                timeLeft = jsonData['time_left']
                print('[OTP Response]', {'jsonData': jsonData, 'message': message, 'timeLeft': timeLeft})
                time.sleep(4)
            except Exception as error:
                print('An error occurred while checking API status:', error)
                break
        print('[Status of the API is]', status)

    signup_button.click()
    time.sleep(5)

    def getRandomInteger(min, max):
        return str(random.randint(min, max))

    month_select = Select(browser.find_element_by_css_selector('select[title="Month:"]'))
    month_select.select_by_value(getRandomInteger(0, 12))

    day_select = Select(browser.find_element_by_css_selector('select[title="Day:"]'))
    day_select.select_by_value(getRandomInteger(0, 29))

    year_select = Select(browser.find_element_by_css_selector('select[title="Year:"]'))
    year_select.select_by_value(getRandomInteger(1965, 2000))
    time.sleep(5)

    next_buttons = browser.find_elements_by_xpath('//button[@type="button"]')
    if len(next_buttons) > 0:
        next_buttons[1].click()

    fetchOTPDetails()
    time.sleep(0.3)
    if status != 1:
        confirmation_code_input = browser.find_element_by_css_selector('input[name=confirmationCode]')
        confirmation_code_input.send_keys(message)
        time.sleep(5)
        confirmation_signup_button = browser.find_elements_by_xpath('//button[@type="button"]')
        if len(confirmation_signup_button) > 0:
            confirmation_signup_button[0].click()
            addAccountDataToProfile()

    browser.close()
